File: hidost/feat_extract.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# feat_extract.py
# Copyright 2014 Nedim Srndic, University of Tuebingen
#
# This file is part of Hidost.
# Hidost is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# Hidost is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Hidost. If not, see <http://www.gnu.org/licenses/>.
#
# Created on December 3, 2014.
"""
This script extracts structural features from SWF files and
saves them in a text file in LibSVM format.
"""
from __future__ import print_function

from argparse import ArgumentParser
import ast
import collections
import multiprocessing
import os
import pickle
import re
import shelve
import subprocess
import sys
import types


def append_libsvm(X, y, fd, comment=[]):
    """
    Appends the data in X and y in LibSVM format to open file
    descriptor fd. Optional comment is appended to the end of
    the line.
    """
    fd.write('{} '.format(int(y)))
    for i, v in enumerate(X, start=1):
        if v != 0.0:
            fd.write('{}:{} '.format(i, v))
    if comment:
        fd.write('#{}'.format(comment))
    fd.write('\n')


def run_SWFExtractor(swffile):
    """
    Runs a child SWFExtractor process on a given file and returns the
    standard output as a string. If the process doesn't terminate in 11
    seconds, it gets killed.
    """
    classpath = '{}:{}'.format(os.path.dirname(__file__),
                               os.getenv('CLASSPATH', ''))
    extractor = subprocess.Popen(['timeout', '-k 1s', '40s',
                                  'java', '-cp', classpath,
                                  'SWFExtractor', swffile],
                                 stdout=subprocess.PIPE)
    (stdoutdata, stderrdata) = extractor.communicate()
    if stderrdata is not None and len(stderrdata) > 0:
        sys.stderr.write('Child (SWFExtractor) error: %s\n' % stderrdata)
        raise
    else:
        return stdoutdata

# Precompiled REs for compacting repetitive SWF structural paths
_comp1_re = re.compile(r'(DefineSprite\x00ControlTags\x00){2,}')
_comp2_re = re.compile(r'(Symbol\x00Name\x00){2,}')
# Precompiled RE for matching a line of SWFExtractor.java output
_line_re = re.compile(r'^(?P<level> *)\[.*?\]: (?P<label>\w+)( : (?P<value>\d+|true|false|[-+]?[0-9]*\.?[0-9]+|.*)|.*)?$')


def get_swf_structure(f, verbose=False, do_compact=True):
    """
    Returns a dictionary containing all structural paths and their
    values in the given SWF file. If verbose, will write a lot to
    sys.stderr. String values are discarded. Returns None on failure.
    """
    swfstr = 0
    try:
        swfstr = run_SWFExtractor(f).splitlines()
    except:
        if verbose:
            sys.stderr.write('Child process exited unexpectedly.\n')
        return None

    if not swfstr or swfstr[-1] != 'OKAY - DONE!':
        if verbose:
            sys.stderr.write('Child process exited unexpectedly.\n')
        if verbose and len(swfstr) >= 1:
            sys.stderr.write(swfstr[-1] + '\n')
        if verbose and len(swfstr) >= 2:
            sys.stderr.write(swfstr[-2] + '\n')
        return None

    s = {}  # SWF file structure
    path_maker = [''] * 1000
    level = 0
    for linenum, line in enumerate(swfstr, start=1):
        if line.find('[') == -1:
            if verbose:
                sys.stderr.write('-{:>03d}: {}\n'.format(linenum, line))
            continue
        if verbose:
            sys.stderr.write(' {:>03d}: {:<60s}'.format(linenum, line))
        match = _line_re.match(line)
        if not match:
            if verbose:
                sys.stderr.write('No match.\n')
            continue
        match = match.groupdict()
        level = len(match['level']) / 2
        path_maker[level] = match['label']
        if match['value'] is None:
            if verbose:
                sys.stderr.write('No value.\n')
        else:
            pathstr = '{}\0'.format('\0'.join(path_maker[:level + 1]))
            if do_compact:
                pathstr = _comp1_re.sub('DefineSprite\x00ControlTags\x00',
                                        pathstr)
                pathstr = _comp2_re.sub('Symbol\x00Name\x00', pathstr)

            # Deduce value type
            try:
                value = ast.literal_eval(match['value'])
            except:
                value = match['value']
                if value == 'true':
                    value = True
                elif value == 'false':
                    value = False

            # Discard strings, Nones, tuples and longs
            if (isinstance(value, str) or
                    isinstance(value, types.NoneType) or
                    isinstance(value, tuple) or
                    isinstance(value, long)):
                continue

            # Save value
            if pathstr in s:
                s[pathstr].append(value)
            else:
                s[pathstr] = [value]

            if verbose:
                sys.stderr.write('{}\n'.format(repr((match['value'], value))))

    # Convert all values of individual paths to the majority type
    for k in s.iterkeys():
        v = s[k]
        if len(v) == 1:
            continue
        c = collections.Counter(map(type, v))
        if len(c) == 1:
            continue
        else:
            try:
                s[k] = map(c.most_common(1)[0][0], v)
            except ValueError as e:
                sys.stderr.write('==> ValueError in get_swf_structure()\n')
                raise e

    return s if s else None


def reduce_structure(f, label, verbose=False):
    """
    Reduces the SWF file structure by discarding information.
    All values belonging to a single path are processed according to
    their type. For integers and floats, all values except for the
    median are discarded. Booleans are reduced to their mean value.
    All returned values are converted to floating-point.
    """
    try:
        s = get_swf_structure(f)
    except:
        return None, label
    if s is None:
        # Probably an error occured in Java code
        return None, label
    for k in s.keys():
        vals = s[k]
        if verbose:
            sys.stderr.write('[{}]: {} '.format(k, repr(vals)))
        if isinstance(vals[0], bool):
            # For booleans, only keep mean
            s[k] = float(sum(vals)) / len(vals)
        elif isinstance(vals[0], int) or isinstance(vals[0], float):
            # For integers and floats, only keep median
            vl = len(vals)
            if vl % 2 == 0:
                s[k] = sum(sorted(vals)[vl / 2 - 1: vl / 2 + 1]) / 2.0
            else:
                s[k] = float(sorted(vals)[vl / 2])
        else:
            raise TypeError('Unexpected value type: {}'.format(type(vals[0])))
        if verbose:
            sys.stderr.write('{}\n'.format(s[k]))
    return s, label


def reduce_structure_fn(args):
    """
    Helper function for multiprocessing.
    """
    return reduce_structure(*args)


def extract_structures(bfs, mfs, cache, N_parallel, verbose=False):
    """
    Extracts and returns SWF file structures from files in bfs and
    mfs. Only processes files missing from cache and caches the
    obtained structures. Uses N_parallel processes.
    """
    pool_args = []

    def check_cache(flist, label):
        for i, f in enumerate(flist, start=1):
            if f not in cache:
                pool_args.append((f, label))

    check_cache(bfs, False)
    check_cache(mfs, True)
    if pool_args:
        print('Files to process (not in cache):', len(pool_args))
        pool = multiprocessing.Pool(N_parallel, maxtasksperchild=10)
        for i, (s, label) in enumerate(pool.imap(reduce_structure_fn,
                                                 pool_args)):
            fname = pool_args[i][0]
            if s is not None:
                s['f'] = fname
                s['m'] = label
                print('{:>5}: {}'.format(i, fname))
            else:
                print('ERROR {:>5}: {}'.format(i, fname))
            cache[fname] = s
        pool.close()
        pool.join()
    else:
        print('All files already cached')

    return [cache[f] for f in bfs + mfs if cache[f] is not None]


def main():
    cachef = os.path.abspath(os.path.join(os.path.dirname(__file__),
                                          'swf.cache'))
    parser = ArgumentParser(description=__doc__)
    parser.add_argument('-b', '--ben',
                        default=False,
                        help='A text file containing a list of paths to '
                        'benign SWF files, one per line.')
    parser.add_argument('-m', '--mal',
                        default=False,
                        help='A text file containing a list of paths to '
                        'malicious SWF files, one per line.')
    parser.add_argument('-o', '--out',
                        required=True,
                        help='Where to save output data file in LibSVM '
                        'format.')
    parser.add_argument('-f', '--features',
                        default=False,
                        help='A pickled list of features to extract; if '
                        'not provided all features found in the dataset '
                        'will be used.')
    parser.add_argument('-s', '--save-features',
                        default=False,
                        help='Where to save a pickled list of all features '
                        'found in the dataset.')
    parser.add_argument('-N', '--N-parallel',
                        type=int, default=1,
                        help='Number of parallel processes to run for '
                        'SWF structure extraction; expect occasional '
                        'explosion of memory use (jumps to >2GB per '
                        'process) due to invoked Java code (default: 1)')
    parser.add_argument('-c', '--cache',
                        default=False,
                        help='Where to cache extracted structures (file '
                        'name, default: {})'.format(cachef))

    args = parser.parse_args()
    bfs, mfs = [], []
    if args.ben:
        bfs = open(args.ben, 'rb').read().splitlines()
    if args.mal:
        mfs = open(args.mal, 'rb').read().splitlines()
    if not args.ben and not args.mal:
        print('At least one of -b, -m is required.')
        return 1

    if args.cache:
        cachef = args.cache
    print('Opening cache file [{}]'.format(cachef))
    cache = shelve.open(cachef)

    print('Extracting uncached SWF file structures')
    structs = extract_structures(bfs, mfs, cache, args.N_parallel)
    print('Successfully extracted {}/{} structures'.format(len(structs),
                                                           len(bfs) +
                                                           len(mfs)))
    feats = 0
    if args.features:
        feats = pickle.load(open(args.features, 'rb'))
    else:
        print('Enumerating all features in the dataset')
        feats = set()
        for s in structs:
            feats.update(s.keys())
        feats = sorted(list(feats.difference(['m', 'f'])))
        if args.save_features:
            print('Saving pickled feature list [{}]'.format(
                args.save_features))
            pickle.dump(feats, open(args.save_features, 'wb+'))
    print('Number of features:', len(feats))

    print('Generating data file [{}]'.format(args.out))
    with open(args.out, 'wb+') as of:
        for s in structs:
            append_libsvm([s.get(p, 0.0) for p in feats],
                          s['m'], of, s['f'])

    return 0

if __name__ == '__main__':
    sys.exit(main())

File: src/feat-extract.cpp
/*
 * Copyright 2014 Nedim Srndic, University of Tuebingen
 *
 * This file is part of Hidost.

 * Hidost is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Hidost is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Hidost.  If not, see <http://www.gnu.org/licenses/>.
 *
 * feat-extract.cpp
 *  Created on: Dec 10, 2013
 */

#include <cstdlib>
#include <fstream>
#include <iostream>
#include <set>
#include <sstream>
#include <string>
#include <vector>

#include <boost/program_options.hpp>
#include <boost/thread.hpp>	// boost::mutex
#include <quickly/DataAction.h>
#include <quickly/ThreadPool.h>

#include "NPPFFile.h"
#include "pdfpath.h"

namespace po = boost::program_options;

typedef std::vector<std::pair<std::string, bool> > filevector;

class DataActionImpl: public quickly::DataActionBase {
private:
    explicit DataActionImpl(unsigned int id) :
            DataActionBase(id) {
    }
    // A mutex for thread safety
    static boost::mutex mutex;
    // Full list of files to process
    static filevector all_files;
    // A set of all paths
    static std::set<std::string> features;
    // The output file
    static std::ofstream out_file;
    // Use values as features
    static bool use_values;

    bool process(const std::string &line);
public:
    // Dummy constructor
    DataActionImpl() :
            DataActionBase(UINT_MAX) {
    }
    virtual ~DataActionImpl() {
    }
    // Virtual constructor
    virtual DataActionImpl *create(unsigned int id) {
        return new DataActionImpl(id);
    }

    // Static constructor
    static void init(const std::string &nppf_name,
                     const std::string &out_file,
                     filevector all_files,
                     bool use_values);

    // Overridden doFull() method
    virtual void doFull(std::stringstream &databuf);
};

boost::mutex DataActionImpl::mutex;
filevector DataActionImpl::all_files;
std::set<std::string> DataActionImpl::features;
std::ofstream DataActionImpl::out_file;
bool DataActionImpl::use_values;

void DataActionImpl::init(const std::string &nppf_name,
                          const std::string &out_file,
                          filevector all_files,
                          bool use_values) {
    for (const auto &feat : InNPPFFile(nppf_name.c_str())) {
        DataActionImpl::features.insert(feat);
    }
    DataActionImpl::out_file.open(out_file, std::ios::binary | std::ios::trunc);
    DataActionImpl::all_files = all_files;
    DataActionImpl::use_values = use_values;
}

bool DataActionImpl::process(const std::string &line) {
    if (line.size() > 0) {
    boost::mutex::scoped_lock lock(DataActionImpl::mutex);
        DataActionImpl::out_file << line << std::endl;
    }
    return true;
}

void DataActionImpl::doFull(std::stringstream &databuf) {
    std::set<std::string>::const_iterator fi = DataActionImpl::features.begin();
    std::string path;
    std::stringstream ss;
    // Write file class
    ss << DataActionImpl::all_files[getId()].second << ' ';
    // Parse paths
    while (databuf.good() and databuf.peek() != EOF) {
        path = get_pdfpath_string(databuf);
        // Remove the space delimiter
        databuf.get();
        // Get the path value
        double val;
        databuf >> val;
        // Remove the trailing space
        databuf.get();
        if (path < *fi) {
            continue;
        } else if (path > *fi) {
            do {
                fi++;
                if (fi == features.end()) {
                    goto end_of_features;
                }
            } while (path > *fi);
        }
        if (path == *fi) {
            // Write feature
            ss << (std::distance(features.begin(), fi) + 1U) << ':';
            if (use_values) {
                ss << val << ' ';
            } else {
                ss << "1 ";
            }
            fi++;
            if (fi == features.end()) {
                goto end_of_features;
            }
        }
    }
    end_of_features:

    // We have to output the feature description for a sample
    // even if not any feature occurs.
    if (DataActionImpl::all_files[getId()].first != "") {
        if (no_features == true) {
            ss << "1:0 ";
            no_features = false;
        }
    }

    // Write file name as comment
    ss << '#' << DataActionImpl::all_files[getId()].first;
    // Do not write empty lines
    while (not this->process(ss.str())) {
    }
}

po::variables_map parse_arguments(int argc, char *argv[]) {
    po::options_description desc(
            "This program extracts PDF structural features from cached "
            "files specified in the input file according to the feature "
            "(NPPF) file and stores them in libsvm format in the output "
            "file. Allowed options");
    desc.add_options()
            ("help", "produce help message")
            ("input-mal,m",
                    po::value<std::string>()->required(),
                    "a list of malicious path files, one per line")
            ("input-ben,b",
                    po::value<std::string>()->required(),
                    "a list of benign path files, one per line")
            ("features,f",
                    po::value<std::string>()->required(),
                    "an NPPF file containing the list of features to extract")
            ("values", "use values instead of presence as features")
            ("vm-limit,M",
                    po::value<unsigned int>()->default_value(0U),
                    "limit the virtual memory of child processes in MB "
                    "(default: no limit)")
            ("cpu-time,t",
                    po::value<unsigned int>()->default_value(0U),
                    "limit the CPU time of child processes in seconds "
                    "(default: no limit)")
            ("output-file,o",
                    po::value<std::string>()->required(),
                    "the feature file to be created")
            ("parallel,N",
                    po::value<unsigned int>()->default_value(0U),
                    "number of child processes to run in parallel "
                    "(default: number of cores minus one)");

    po::variables_map vm;
    po::store(po::command_line_parser(argc, argv).options(desc).run(), vm);

    if (vm.count("help")) {
        std::cout << desc << std::endl;
        std::exit(EXIT_SUCCESS);
    }

    try {
        po::notify(vm);
    } catch (std::exception &e) {
        std::cerr << e.what() << std::endl << std::endl << desc << std::endl;
        std::exit(EXIT_FAILURE);
    }
    return vm;
}

void get_input_files(filevector &input_files, const char *in_name,
                     bool data_class) {
    std::string line;
    std::ifstream ifile(in_name, std::ios::binary);
    while (std::getline(ifile, line)) {
        input_files.push_back(std::pair<std::string, bool>(line, data_class));
    }
    ifile.close();
}

int run(int argc, char *argv[]) {
    // Parse arguments
    po::variables_map vm = parse_arguments(argc, argv);
    const std::string INPUT_MAL = vm["input-mal"].as<std::string>();
    const std::string INPUT_BEN = vm["input-ben"].as<std::string>();
    const std::string NPPF_FILE = vm["features"].as<std::string>();
    const bool USE_VALUES = vm.count("values") > 0;
    const std::string OUTPUT_FILE = vm["output-file"].as<std::string>();
    const unsigned int VM_LIMIT = vm["vm-limit"].as<unsigned int>();
    const unsigned int CPU_LIMIT = vm["cpu-time"].as<unsigned int>();
    const unsigned int PARALLEL = vm["parallel"].as<unsigned int>();

    // Read file list
    filevector input_files;
    get_input_files(input_files, INPUT_MAL.c_str(), true);
    get_input_files(input_files, INPUT_BEN.c_str(), false);
    DataActionImpl::init(NPPF_FILE, OUTPUT_FILE, input_files, USE_VALUES);

    // Construct a vector of command-line arguments
    std::vector<const char * const *> argvs;
    const char prog_name[] = "/bin/cat";
    const char *pn = prog_name;
    for (const auto &file : input_files) {
        const char **argv = (const char **) NULL;
        try {
            argv = new const char *[3]{pn, file.first.c_str(), nullptr};
        } catch (...) {
            delete[] argv;
        }
        argvs.push_back(argv);
    }

    // Prepare the data action and perform scan
    DataActionImpl dummy;
    quickly::ThreadPool pool(pn, argvs, &dummy, PARALLEL);
    pool.setVerbosity(5U);
    pool.setVmLimit(VM_LIMIT * 1024U * 1024U);
    pool.setCpuLimit(CPU_LIMIT);
    pool.run();

    // Print results, delete command-line arguments
    for (unsigned int i = 0U; i < argvs.size(); i++) {
        delete[] argvs[i];
    }

    return EXIT_SUCCESS;
}

int main(int argc, char *argv[]) {
	try {
		return run(argc, argv);
	} catch (std::exception &e) {
		std::cerr << "Exception caught: " << e.what() << std::endl;
		return EXIT_FAILURE;
	} catch (const char *e) {
		std::cerr << "Exception caught: " << e << std::endl;
		return EXIT_FAILURE;
	} catch (...) {
		std::cerr << "Unexpected exception caught." << std::endl;
		return EXIT_FAILURE;
	}
}
